esp/platform/txsummary.cpp

```diff
@@ -65,100 +398,490 @@ bool CTxSummary::contains(const char* key) const
+
+    return true;
+}
+
+// Set a new entry named 'destKey' with it's value copied from the entry named 'sourceKey'
+// Group and LogLevel of the new entry are set as passed in.
+bool CTxSummary::setCopyValueOf(const char* destKey, const char* sourceKey, const LogLevel logLevel, const unsigned int group)
+{
+    if(!validateKey(destKey))
+        return false;
+
+    CriticalBlock block(m_sync);
+
+    if(!__contains(sourceKey))
+        return false;
+
+    TxEntryBase* sourceEntry = queryEntry(sourceKey);
+    Owned<TxEntryBase> newEntry = sourceEntry->cloneWithNewName(destKey);
+    newEntry->logLevel = logLevel;
+    newEntry->group = group;
+    set(destKey, newEntry.getLink());
+    return true;
+}
+
+
+CumulativeTimer* CTxSummary::queryTimer(const char* key)
+{
+    if(!validateKey(key))
+        return nullptr;
+
+    CriticalBlock block(m_sync);
+
+    Linked<TxEntryBase> entry = queryEntry(key);
+
+    if(entry)
     {
-        timer.setown(new CumulativeTimer());
+        // Is the entry a Cumulative Timer?
+        TxEntryTimer* timer = dynamic_cast<TxEntryTimer*>(entry.get());
+        if(!timer)
+            return nullptr;
+        else
+            return timer->value.get();
+    } else {
+        // The CumulativeTimer does not yet exist
+        // create it and add to the list of entries
+        Owned<CumulativeTimer> newTimer = new CumulativeTimer();
+        EntryValue newEntry;
+        newEntry.setown(new TxEntryTimer(key, newTimer.get()));
+        m_entries2.push_back(newEntry);
+        return newTimer.get();
     }
-
-    return timer;
 }
 
-bool CTxSummary::updateTimer(const char* name, unsigned long long delta, const LogLevel logLevel)
+bool CTxSummary::updateTimer(const char* name, unsigned long long delta, const LogLevel logLevel, const unsigned int group)
 {
     Owned<CumulativeTimer> timer;
 
+    CriticalBlock block(m_sync);
+
     timer.set(queryTimer(name));
 
     if (!timer)
         return false;
 
     timer->add(delta);
     timer->setLogLevel(logLevel);
+    timer->setGroup(group);
     return true;
 }
 
-void CTxSummary::serialize(StringBuffer& buffer, const LogLevel logLevel) const
+void CTxSummary::serialize(StringBuffer& buffer, ITxSummaryProfile* profile, const LogLevel logLevel, const unsigned int group, const unsigned int requestedStyle) const
 {
     CriticalBlock block(m_sync);
 
-    for (const Entry& entry : m_entries)
+    if(requestedStyle & TXSUMMARY_OUT_TEXT)
     {
-        if (entry.logLevel > logLevel)
-            continue;
-
-        if (entry.value.length())
-            buffer.appendf("%s=%s;", entry.key.str(), entry.value.str());
-        else
-            buffer.appendf("%s;", entry.key.str());
+        // For text serialization, we create a new entry if it has
+        // a profile mapping it to a new name, then serialize it.
+        // Otherwise, we just serialize the existing entry.
+        for(auto entry : m_entries2)
+        {
```

Comment:
It might be cleaner for most of this block to exist in entry. If the entry determines it needs to be renamed while serializing itself, it can rename itself in output without cloning.

Reply:
I agree it sounds nice, but then I'd need to pass down CTxSummary into Entry::serialize to get at the profile and also __contains(). I think I'll leave as-is.